### PR TITLE
Add protocol buffer language output

### PR DESCRIPTION
**Pull request by jhalderm**
This PR adds protocol buffer support to ZSchema.

### CODE REVIEW

File: zschema/compounds.py [hunk 1]
```diff
@@ -209,6 +224,27 @@ def to_bigquery(self, name):
...
+
+        global _proto_messages
+        if anon or message_type not in _proto_messages:
+            fields = [v.to_proto(k, indent) for k,v in \
+                      sorted(self.definition.iteritems(), key=lambda (k,v): v.sort_index)]
+            fields = ["%s = %d;" % (fields[i], i+1) for i in range(len(fields))]
+            proto_def = "message %s {\n%s\n}" % (message_type, _proto_indent("\n".join(fields), indent+1))
+            _proto_messages[message_type] = True
+            if not anon:
+                print proto_def
```
Comment (zakird) [inline zschema/compounds.py]: Is this ended to be a print here?

**Comment by zakird** [comment]
Would you be willing to add a couple of simple tests that to show the expected output?

File: zschema/compounds.py (hunk 1 above)
Comment (jhalderm) [reply zschema/compounds.py]: Yes.  It's emitting a message definition at the global level, rather than an embedded anonymous message definition.

**Comment by jhalderm** [comment]
@zakird Test added.

**Comment by jhalderm** [comment]
Commit above restructures the patch to fix a logic error with lists of anonymous subrecords, and to avoid printing internally.  The output for ZTag's `ipv4host` schema now compiles with protoc.

File: zschema/compounds.py [hunk 2]
```diff
@@ -10,6 +11,18 @@ def _is_valid_object(name, object_):
     if not isinstance(object_, Keyable):
         raise Exception("Invalid schema. %s is not a Keyable." % name)
 
+def _proto_message_name(string):
+    if string != string.lower():
+        return string
+    string = "".join(w.capitalize() for w in string.split("_"))
+    string = string.replace("Ztag", "ZTag")
```
Comment (zakird) [inline zschema/compounds.py]: This scares me a bit. Where does ZTag show up in current schemas? Is it only ZTag? Does ZGrab show up?

**Comment by zakird** [comment]
Overall this looks good to me, but I don't think I understand based on this logic how the code keeps record IDs consistent across multiple runs. If other fields are later added, does anything prevent the previously ID'ed fields from changing ID?

**Comment by jhalderm** [comment]
Every time a keyable is created, is gets assigned a value from an incrementing global index.  The fields in a proto message are sorted based on these values, then assigned a local index.

This means the IDs will be stable so long as fields are never removed and new fields are only added at the end of a structure.

File: zschema/compounds.py (hunk 2 above)
Comment (jhalderm) [reply zschema/compounds.py]: I'm going to remove this.

**Comment by zakird** [comment]
Hmm. That makes a lot of sense. It's a definite trade off of "if someone accidentally moves something around to be human readable it'll break our protobufs" vs "being (painfully) explicit". I'm not sure the right call. Not being able to rearrange seems like it _could_ cause problems, but being explicit in the schema is also somewhat annoying.

**Comment by jhalderm** [comment]
Added support for explicit field indices.  If you specify `pr_index=n` on a leaf, it will be assigned that index in the proto definition.  The patch prohibits mixing implicit and explicit indices within a subrecord, since that's unlikely to be what the programmer wants.

**Comment by jhalderm** [comment]
Last commit removes the ugly exception handler.  I should have just set a default in `SubRecord`.

**Comment by jhalderm** [comment]
I've added a property, `pr_omit`, that  you can set to have a keyable ignored by the protocol buffer conversion.

**Comment by jhalderm** [comment]
@zakird, would appreciate another look from you.  If you're okay with this, we can merge.

**Comment by zakird** [comment]
Apologies, I was hiking for a few days this week and didn't have a chance to look at this PR until tonight. Overall this looks good, and I'm happy with the merge. I'm a bit nervous that `_proto_messages` is a global, which could cause issues in library-mode, but we can tackle that when it comes up.

**Pull request merged**